src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBuildReferenceTaxonomyUtils.java

```diff
@@ -0,0 +1,318 @@
+    /**
+     * Helper function for building the map from tax id to reference contig accession
+     */
+    private static void updateAccessionTaxonProperties(final String accession, final String name, final long length, final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties) {
+        final PSPathogenReferenceTaxonProperties taxonProperties;
+        if (taxIdToProperties.containsKey(accession)) {
+            taxonProperties = taxIdToProperties.get(accession);
+        } else {
+            taxonProperties = new PSPathogenReferenceTaxonProperties();
+        }
+        taxonProperties.accessions.add(name);
+        taxonProperties.length += length;
+        taxIdToProperties.put(accession, taxonProperties);
+    }
+
+    /**
+     * Create reference_name-to-taxid map (just an inversion on taxIdToProperties)
+     */
+    protected static Map<String, String> buildAccessionToTaxIdMap(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties) {
+        final Map<String, String> accessionToTaxId = new HashMap<>();
+        for (final String taxId : taxIdToProperties.keySet()) {
+            final PSPathogenReferenceTaxonProperties taxonProperties = taxIdToProperties.get(taxId);
+            for (String name : taxonProperties.accessions) {
+                accessionToTaxId.put(name, taxId);
+            }
+        }
+        return accessionToTaxId;
+    }
+
+    /**
+     * Returns a PSTree representing a reduced taxonomic tree containing only taxa present in the reference
+     */
+    protected static PSTree buildTaxonomicTree(final Map<String, PSPathogenReferenceTaxonProperties> taxIdToProperties) {
+
+        //Build tree of all taxa
+        final PSTree tree = new PSTree(ROOT_ID);
+        final Collection<String> invalidIds = new ArrayList<>();
+        for (final String taxId : taxIdToProperties.keySet()) {
+            if (!taxId.equals(ROOT_ID)) {
+                final PSPathogenReferenceTaxonProperties taxonProperties = taxIdToProperties.get(taxId);
+                if (taxonProperties.name != null && taxonProperties.parentTaxId != null && taxonProperties.rank != null) {
+                    tree.addNode(taxId, taxonProperties.name, taxonProperties.parentTaxId, taxonProperties.length, taxonProperties.rank);
+                } else {
+                    invalidIds.add(taxId);
+                }
+            }
+        }
+        PSUtils.logItemizedWarning(logger, invalidIds, "The following taxonomic IDs did not have name/taxonomy information (this may happen when the catalog and taxdump files are inconsistent)");
+
+        final Set<String> unreachableNodes = tree.removeUnreachableNodes();
+        if (!unreachableNodes.isEmpty()) {
+            PSUtils.logItemizedWarning(logger, unreachableNodes, "Removed " + unreachableNodes.size() + " unreachable tree nodes");
+        }
+
+        tree.checkStructure();
+
+        //Trim tree down to nodes corresponding only to reference taxa (and their ancestors)
```

Comment:
I don't think so.. for each node that has a reference, you add it and all its ancestors to relevantNodes. If you do that while constructing the tree, you're not guaranteed that all its ancestors have been added to the tree yet.